script/windows/14.0/update/WindowsUpdateCheck.php

```diff
@@ -0,0 +1,164 @@
+
+  class WindowsUpdateCheck {
+    const MSI_REGEX = '/^keyman(desktop)?\.msi$/';
+    const BOOTSTRAP_REGEX = '/^setup\.exe$/';
+    const BUNDLE_REGEX = '/^keyman(desktop)?-.+\.exe/';
+
+    public function execute($tier, $appVersion, $packages) {
+      $desktop_update = [];
+
+      $desktop_update['msi'] = $this->BuildKeymanDesktopVersionResponse($tier, $appVersion, self::MSI_REGEX);
+      $desktop_update['setup'] = $this->BuildKeymanDesktopVersionResponse($tier, $appVersion, self::BOOTSTRAP_REGEX);
+      $desktop_update['bundle'] = $this->BuildKeymanDesktopVersionResponse($tier, $appVersion, self::BUNDLE_REGEX);
+      if(!empty($desktop_update['bundle'])) {
+        $newAppVersion = $desktop_update['bundle']->version;
+      } else {
+        $newAppVersion = $appVersion;
+      }
+      $desktop_update['keyboards'] = $this->BuildKeyboardsResponse($newAppVersion, $packages);
+
+      return json_encode($desktop_update, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+    }
+
+    private function BuildKeymanDesktopVersionResponse($tier, $InstalledVersion, $regex) {
+      if(empty($this->DownloadVersions)) {
+        $this->DownloadVersions = @file_get_contents(get_site_url_downloads() . "/api/version/windows/2.0");
+        if($this->DownloadVersions === FALSE) {
+          fail('Unable to retrieve version data from '.get_site_url_downloads(), 500);
+        }
+        $this->DownloadVersions = @json_decode($this->DownloadVersions);
+        if($this->DownloadVersions === NULL) {
+          fail('Unable to decode version data from '.get_site_url_downloads(), 500);
+        }
+
+        if(!isset($this->DownloadVersions->windows)) {
+          fail("Unable to find {windows} key in ".get_site_url_downloads()." data", 500);
+        }
+      }
+
+      // Check each of the tiers for the one that matches the major version.
+      // This gets us upgrades on alpha, beta and stable tiers.
+      $tiers = get_object_vars($this->DownloadVersions->windows);
+
+      return $this->CheckVersionResponse($tier, $tiers, $InstalledVersion, $regex);
+    }
+
+    private function CheckVersionResponse($tier, $tiers, $InstalledVersion, $regex) {
+      if(!isset($tiers[$tier])) return FALSE;
+      $tierdata = $tiers[$tier];
+
+      // We will support staying on alpha or beta tier once a version
+      // hits stable. This is correct, with major upgrade support, as
+      // that'll mean the user will transition to a new version on the
+      // same tier (i.e. always staying bleeding edge)
+
+      $files = get_object_vars($tierdata->files);
+      foreach($files as $file => $filedata) {
+        // This is currently tied to Windows -- for other platforms we need to change this
+        if(preg_match($regex, $file)) {
+          $filedata->url = get_site_url_downloads() . "/windows/$tier/{$filedata->version}/{$file}";
+          return $filedata;
+        }
+      }
+
+      return FALSE;
+    }
+
+    private function IsSameMajorVersion($v1, $v2) {
```

Comment:
This looks unused and maybe can be removed

Reply:
Noted for cleanup in follow-up PR.